app/models.py

```diff
@@ -2551,3 +2556,92 @@ def serialize(self):
             "original_filename": self.original_filename,
             "page_count": self.page_count,
         }
+
+
+class UnsubscribeRequestReport(db.Model):
+    __tablename__ = "unsubscribe_request_report"
+    id = db.Column(UUID(as_uuid=True), primary_key=True)
+
+    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey("services.id"), nullable=False)
+    service = db.relationship(Service, backref=db.backref("unsubscribe_request_reports"))
+
+    earliest_timestamp = db.Column(db.DateTime, nullable=False, default=datetime.datetime.utcnow)
+    latest_timestamp = db.Column(db.DateTime, nullable=False, default=datetime.datetime.utcnow)
+    processed_by_service_at = db.Column(db.DateTime, nullable=True)
+    count = db.Column(db.BigInteger, nullable=False)
+
+
+class UnsubscribeRequest(db.Model):
+    __tablename__ = "unsubscribe_request"
+    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
+
+    # WIP:
+    # Ignoring a strict foreign key relationship here for now. Notifications are archived to the NotificationHistory
+    # table by a nightly job and I haven't investigated whether that might break a strict FK yet or if it would
+    # work smoothly. We can still have a relationship using an explicit join condition.
+    notification_id = db.Column(UUID(as_uuid=True), index=True, nullable=False)
+    notification = db.relationship(
+        "NotificationAllTimeView",
+        primaryjoin="UnsubscribeRequest.notification_id == foreign(NotificationAllTimeView.id)",
+        uselist=False,
+    )
+
+    # this is denormalised but might still be useful to have as a separate column?
+    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey("services.id"), nullable=False)
+    service = db.relationship(Service, backref=db.backref("unsubscribe_requests"))
+
+    template_id = db.Column(UUID(as_uuid=True), nullable=False)
+    template_version = db.Column(db.Integer, nullable=False)
+
+    template_history = db.relationship(TemplateHistory, backref=db.backref("unsubscribe_requests"))
+    template = db.relationship(
+        Template,
+        foreign_keys=[template_id],
+        primaryjoin="Template.id == UnsubscribeRequest.template_id",
+        backref=db.backref("unsubscribe_requests"),
+    )
+
+    email_address = db.Column(db.Text, nullable=False)
+
+    created_at = db.Column(db.DateTime, nullable=False, default=datetime.datetime.utcnow)
+
+    # is this useful or is a boolean enough? if we create a table for a report, we could remove from this table entirely
+    processed_at = db.Column(db.DateTime, nullable=True)
```

Comment:
ive actually removed this sorry! i thought about it because for now we're just thinking about reports, and the "what if these timestamps end up different" question was bugging me a bit

Reply:
Fair enough, I don't feel too strongly either way as it seems we have to ship this and make adjustments to the tables if or as needed.